WebKit managed downloads should consult the Screen Time web allow/block list

beidson opened the pull request:
#### 500826177f6a860b62e7270da16f27ed71bc578e
<pre>
WebKit managed downloads should consult the Screen Time web allow/block list
<a href="https://rdar.apple.com/162877418">rdar://162877418</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=303971">https://bugs.webkit.org/show_bug.cgi?id=303971</a>

Reviewed by Per Arne Vollan.

When a navigation occurs, WebKit will automatically subject that navigation to a few different
methods of content filtering defined by the platform.

One such method of filtering is Screen Time&apos;s parental controls based on URL block/allow lists.

When a navigation turns into a download, then further redirects of the URL should continue to
consult the URL block/allow lists.

The same is true for downloads started directly via WKWebView/WKDownload APIs.

Most of the change here is in PendingDownload, which always sits between the actual network load
and clients being informed of download-related decisions and progress.

PendingDownload will now consult its initial URL against the block/allow lists, as well as any redirects.

If it is informed that the URL is blocked, it will cancel the network load and schedule its own
failure with an appropriate error to notify the WebKit API client.

* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:

* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
(WebCore::ParentalControlsURLFilter::setSharedParentalControlsURLFilterIfNecessary):
(WebCore::ParentalControlsURLFilter::isURLAllowed):
(WebCore::ParentalControlsURLFilter::isURLAllowedImpl):

* Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
(WebKit::PendingDownload::PendingDownload):
(WebKit::PendingDownload::willSendRedirectedRequest):
(WebKit::PendingDownload::messageSenderDestinationID const):
(WebKit::PendingDownload::blockDueToContentFilter):
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.h:

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startContentFiltering):
(WebKit::setSharedParentalControlsURLFilterIfNecessary): Deleted.

* Source/WebKit/Shared/ios/WebParentalControlsURLFilter.h:
* Source/WebKit/Shared/ios/WebParentalControlsURLFilter.mm:
(WebKit::WebParentalControlsURLFilter::isURLAllowedImpl):
(WebKit::WebParentalControlsURLFilter::isURLAllowed): Deleted.
(WebKit::WebParentalControlsURLFilter::allowURL): Deleted.

Canonical link: <a href="https://commits.webkit.org/304323@main">https://commits.webkit.org/304323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df9943f8211e832db9af7dfcb61250c787107e0d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135243 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46510 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142750 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137112 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8270 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7482 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138189 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5902 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/121222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84212 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3339 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145446 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7316 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39965 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7360 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112096 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28444 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5538 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117516 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7370 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7126 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/70921 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7346 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7229 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
```diff
@@ -43,22 +44,53 @@ WTF_MAKE_TZONE_ALLOCATED_IMPL(PendingDownload);
+    NetworkProcess::setSharedParentalControlsURLFilterIfNecessary();
+#endif
+
+#if HAVE(WEBCONTENTRESTRICTIONS)
+#if HAVE(WEBCONTENTRESTRICTIONS_PATH_SPI)
+    m_urlFilter = ParentalControlsURLFilter::filterWithConfigurationPath();
+#else
+    m_urlFilter = ParentalControlsURLFilter::singleton();
+#endif // HAVE(WEBCONTENTRESTRICTIONS_PATH_SPI)
+#endif // HAVE(WEBCONTENTRESTRICTIONS)
```
I don't think we need to address this in this patch, but should downloads be checked against all content filters?

beidson replied:
One the one hand, it seems like it would be desirable (And my first attempt at this did, in fact, try to run it against ContentFilter and all implementations)

A few of us who discussed concluded:
1 - Actual content analysis of the downloads is not possible in WebKit today, at least in the Cocoa port, because we never see the bytes
2 - Our actual content analysis frameworks that WebKit uses are very targeted to small bits of web content, and not a 100mb PDF file or a 2gb media file, e.g.
3 - Consulting the URL filters solves the task at hand
4 - If we decide we need to analyze the entire file later, there's other strategies that might make way more sense than having WebKit do it at download time.

pvollan commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -288,25 +288,14 @@ void NetworkResourceLoader::startRequest(const ResourceRequest& newRequest)
         return;
     }
-    setSharedParentalControlsURLFilterIfNecessary();
+    NetworkProcess::setSharedParentalControlsURLFilterIfNecessary();
```
We could potentially call this on startup of the Network process, since we will always call this method, IIUC.

beidson replied:
I definitely agree in principle.

In the interests of this patch being as small of a behavior change as possible, I'll leave it as is.

I'll do that in a followup.

beidson committed:
WebKit managed downloads should consult the Screen Time web allow/block list
rdar://162877418
https://bugs.webkit.org/show_bug.cgi?id=303971

Reviewed by Per Arne Vollan.

When a navigation occurs, WebKit will automatically subject that navigation to a few different
methods of content filtering defined by the platform.

One such method of filtering is Screen Time's parental controls based on URL block/allow lists.

When a navigation turns into a download, then further redirects of the URL should continue to
consult the URL block/allow lists.

The same is true for downloads started directly via WKWebView/WKDownload APIs.

Most of the change here is in PendingDownload, which always sits between the actual network load
and clients being informed of download-related decisions and progress.

PendingDownload will now consult its initial URL against the block/allow lists, as well as any redirects.

If it is informed that the URL is blocked, it will cancel the network load and schedule its own
failure with an appropriate error to notify the WebKit API client.

* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:

* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
(WebCore::ParentalControlsURLFilter::setSharedParentalControlsURLFilterIfNecessary):
(WebCore::ParentalControlsURLFilter::isURLAllowed):
(WebCore::ParentalControlsURLFilter::isURLAllowedImpl):

* Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
(WebKit::PendingDownload::PendingDownload):
(WebKit::PendingDownload::willSendRedirectedRequest):
(WebKit::PendingDownload::messageSenderDestinationID const):
(WebKit::PendingDownload::blockDueToContentFilter):
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.h:

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startContentFiltering):
(WebKit::setSharedParentalControlsURLFilterIfNecessary): Deleted.

* Source/WebKit/Shared/ios/WebParentalControlsURLFilter.h:
* Source/WebKit/Shared/ios/WebParentalControlsURLFilter.mm:
(WebKit::WebParentalControlsURLFilter::isURLAllowedImpl):
(WebKit::WebParentalControlsURLFilter::isURLAllowed): Deleted.
(WebKit::WebParentalControlsURLFilter::allowURL): Deleted.

Canonical link: https://commits.webkit.org/304323@main
